builtin/credential/aws/path_login.go

```diff
@@ -0,0 +1,466 @@
+	if storedIdentity != nil {
+		// Check if the client nonce match the cached nonce and if the pending time
+		// of the identity document is not before the pending time of the document
+		// with which previous login was made.
+		if err = validateMetadata(clientNonce, identityDoc.PendingTime, storedIdentity, imageEntry); err != nil {
+			return logical.ErrorResponse(err.Error()), nil
+		}
+	}
+
+	// Load the current values for max TTL and policies from the image entry,
+	// before checking for overriding by the RoleTag
+	maxTTL := b.System().MaxLeaseTTL()
+	if imageEntry.MaxTTL > time.Duration(0) && imageEntry.MaxTTL < maxTTL {
+		maxTTL = imageEntry.MaxTTL
+	}
+
+	policies := imageEntry.Policies
+	rTagMaxTTL := time.Duration(0)
+	disallowReauthentication := imageEntry.DisallowReauthentication
+
+	// Role tag is enabled for the AMI.
+	if imageEntry.RoleTag != "" {
+		// Overwrite the policies with the ones returned from processing the role tag.
+		resp, err := b.handleRoleTagLogin(req.Storage, identityDoc, imageEntry, instanceDesc)
+		if err != nil {
+			return nil, err
+		}
+		policies = resp.Policies
+		rTagMaxTTL = resp.MaxTTL
+
+		// If imageEntry had disallowReauthentication set to 'true', do not reset it
+		// to 'false' based on role tag having it not set. But, if role tag had it set,
+		// be sure to override the value.
+		if !disallowReauthentication {
+			disallowReauthentication = resp.DisallowReauthentication
+		}
+
+		if resp.MaxTTL > time.Duration(0) && resp.MaxTTL < maxTTL {
+			maxTTL = resp.MaxTTL
+		}
+	}
+
+	// Save the login attempt in the identity whitelist.
+	currentTime := time.Now()
+	if storedIdentity == nil {
+		// AmiID, ClientNonce and CreationTime of the identity entry,
+		// once set, should never change.
+		storedIdentity = &whitelistIdentity{
+			AmiID:        identityDoc.AmiID,
+			ClientNonce:  clientNonce,
+			CreationTime: currentTime,
+		}
+	}
+
+	// DisallowReauthentication, PendingTime, LastUpdatedTime and ExpirationTime may change.
+	storedIdentity.LastUpdatedTime = currentTime
+	storedIdentity.ExpirationTime = currentTime.Add(maxTTL)
+	storedIdentity.PendingTime = identityDoc.PendingTime
+	storedIdentity.DisallowReauthentication = disallowReauthentication
+
+	// Performing the clientNonce empty check after determining the DisallowReauthentication
+	// option. This is to make clientNonce optional when DisallowReauthentication is set.
+	if clientNonce == "" && !storedIdentity.DisallowReauthentication {
+		return logical.ErrorResponse("missing nonce"), nil
+	}
+
+	// Limit the nonce to a reasonable length.
+	if len(clientNonce) > 128 && !storedIdentity.DisallowReauthentication {
+		return logical.ErrorResponse("client nonce exceeding the limit of 128 characters"), nil
+	}
+
+	if err = setWhitelistIdentityEntry(req.Storage, identityDoc.InstanceID, storedIdentity); err != nil {
+		return nil, err
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies: policies,
+			Metadata: map[string]string{
```

Comment:
Can we also put the AMI image id in metadata? Might be useful if people want to search audit logs and bulk-revoke.